From 20537d65166714f083256cf4f48b96d35786253a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?G=C3=BCnther=20Deschner?= <gd@samba.org>
Date: Mon, 17 May 2010 18:22:37 +0200
Subject: [PATCH] s3-pidfile: set the close on exec flag for the created pidfiles.

Guenther
---
 source3/lib/pidfile.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/source3/lib/pidfile.c b/source3/lib/pidfile.c
index 2c52d12..a3e39db 100644
--- a/source3/lib/pidfile.c
+++ b/source3/lib/pidfile.c
@@ -145,6 +145,9 @@ void pidfile_create(const char *program_name)
 	}
 	/* Leave pid file open & locked for the duration... */
 	SAFE_FREE(name);
+
+	/* set the close on exec so that we don't leak the fd */
+	fcntl(fd, F_SETFD, FD_CLOEXEC);
 }
 
 void pidfile_unlink(void)
-- 
1.6.6.1

